### PR TITLE
Fix custom block volume migration

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

**Comment by stgraber** [comment]
jenkins not happy

**Comment by monstermunchkin** [comment]
I'm tempted to just drop the second commit. The size thing with LVM doesn't make much sense anyway. Even if I do `size=5MiB` it will create a 8 MiB volume since it uses its physical extents (defaults to 4 MiB) for calculating the size. Therefore, `size` will most likely differ from the actual size.

**Comment by stgraber** [comment]
Hmm, yeah, I guess since it always rounds to upper 4MiB we can just ignore it for now.

File: lxd/storage/drivers/generic_vfs.go [hunk 1]
```diff
@@ -908,3 +908,17 @@ func genericVFSCopyVolume(d Driver, initVolume func(vol Volume) (func(), error),
...
+	if err != nil {
+		return -1, err
+	}
+
+	stat, err := os.Lstat(path)
+	if err != nil {
+		return -1, err
+	}
+
+	return stat.Size(), nil
```
Comment (stgraber) [inline lxd/storage/drivers/generic_vfs.go:923]: This isn't recursive though so would only give you the size of a dir entry, no?

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 2]
```diff
@@ -1821,3 +1829,13 @@ func (d *zfs) RenameVolumeSnapshot(vol Volume, newSnapshotName string, op *opera
 
 	return nil
 }
+
+// GetGetVolumeSize returns a volume's size.
+func (d *zfs) GetVolumeSize(vol Volume) (int64, error) {
+	size, err := d.getDatasetProperty(d.dataset(vol, false), "available")
```
Comment (stgraber) [inline lxd/storage/drivers/driver_zfs_volumes.go:1835]: This doesn't feel like the right properly for this.

If you want the max size, you want quota/refquota (container) or volsize (VM).
If you want the currently used size, then you want referenced (container) or volsize (VM).

File: lxd/storage/drivers/driver_ceph_utils.go [hunk 3]
```diff
@@ -1200,3 +1200,25 @@ func (d *ceph) receiveVolume(volumeName string, conn io.ReadWriteCloser, writeWr
...
 	return nil
 }
+
+func (d *ceph) rbdGetVolumeSize(vol Volume) (string, error) {
+	msg, err := shared.RunCommand(
+		"rbd",
+		"--id", d.config["ceph.user.name"],
+		"--cluster", d.config["ceph.cluster_name"],
+		"--pool", d.config["ceph.osd.pool_name"],
+		"info",
```
Comment (stgraber) [inline lxd/storage/drivers/driver_ceph_utils.go:1210]: Doesn't this one have an option to get json output?

File: lxd/storage/drivers/generic_vfs.go (hunk 1 above)
Comment (monstermunchkin) [reply lxd/storage/drivers/generic_vfs.go:923]: I just checked the output, and it returns the correct size.

File: lxd/storage/backend_lxd.go [hunk 4]
```diff
@@ -2387,6 +2387,22 @@ func (b *lxdBackend) CreateCustomVolumeFromCopy(projectName string, volName stri
 		return fmt.Errorf("Failed to negotiate copy migration type: %v", err)
 	}
 
+	// If we're copying block volumes, the target block volume needs to be
+	// at least the size of the source volume, otherwise we'll run into
+	// "no space left on device".
+	var volSize int64
+
+	if contentType == drivers.ContentTypeBlock {
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:2395]: @monstermunchkin I'm coming into this a bit cold, but is there a reason you couldn't use the same approach we used for block VM root disks here? https://github.com/lxc/lxd/blob/master/lxd/storage/backend_lxd.go#L702-L708

I.e. using drivers.BlockDiskSizeBytes(rootDrivePath)?

Comment (monstermunchkin) [reply lxd/storage/backend_lxd.go:2395]: That decreases the diff quite a bit ;)

**Comment by stgraber** [comment]
As this is storage related, I want to see this 100% green.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
@tomponline can you review?

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 5]
```diff
@@ -600,6 +601,13 @@ func (d *zfs) CreateVolumeFromCopy(vol Volume, srcVol Volume, copySnapshots bool
 func (d *zfs) CreateVolumeFromMigration(vol Volume, conn io.ReadWriteCloser, volTargetArgs migration.VolumeTargetArgs, preFiller *VolumeFiller, op *operations.Operation) error {
 	// Handle simple rsync and block_and_rsync through generic.
 	if volTargetArgs.MigrationType.FSType == migration.MigrationFSType_RSYNC || volTargetArgs.MigrationType.FSType == migration.MigrationFSType_BLOCK_AND_RSYNC {
+		// Round up the volume size to avoid running into "no space left on device".
```
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:604]: @monstermunchkin is this part still needed now you're reading the exact size from the source?

I'm wondering why we didn't encounter issues with this when migrating instance block volumes.

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 6]
```diff
@@ -600,6 +601,13 @@ func (d *zfs) CreateVolumeFromCopy(vol Volume, srcVol Volume, copySnapshots bool
...
+		if volTargetArgs.VolumeSize > 0 {
+			volSizeBytes := math.Ceil(float64(volTargetArgs.VolumeSize) / MinBlockBoundary * MinBlockBoundary)
```
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:606]: The ZFS `createVolume()` function already rounds anyway https://github.com/lxc/lxd/blob/master/lxd/storage/drivers/driver_zfs_utils.go#L58, so if there is a fix that is needed it should probably be in there.

This is used by `CreateVolume()` which is called by `genericVFSCreateVolumeFromMigration`

**Comment by monstermunchkin** [comment]
@tomponline care to take another look?

**Comment by tomponline** [comment]
@monstermunchkin although looks like you've not got your editor applying gofmt on save as code style tests are failing.

**Comment by stgraber** [comment]
jenkins: test this please

**Pull request merged**